src/agreementManager/AgreementManager.ts

```diff
@@ -104,37 +112,293 @@ class AgreementManager {
         let virtualVotingBlocks: BlockConfirmationStruct[] = [];
-        let requiredSignatures = new Set<Address>(thresholdAddresses);
+        let milestoneProofs: MilestoneProofStruct[] = [];
+        let milestoneSnapshots: StateSnapshot[] = [];
+
+        // Start with genesis snapshot for the first milestone
+        let currentSnapshot = genesisSnapshot;
+
+        // Process exit points in chronological order - each exit point is a milestone
+        const sortedExitPoints = [...exitPoints].sort(
+            (a, b) => Number(a) - Number(b)
+        );
+
+        for (const exitPointHeight of sortedExitPoints) {
+            const milestoneData = this.processMilestone(
+                exitPointHeight,
+                blockEntries,
+                currentSnapshot
+            );
+
+            if (milestoneData) {
+                milestoneProofs.push(milestoneData.milestoneProof);
+                milestoneSnapshots.push(milestoneData.snapshot);
+                currentSnapshot = milestoneData.snapshot;
+            }
+        }
+
+        // Find latest finalized and correct states using the final participant set
+        const finalParticipantSet = new Set<Address>(
+            currentSnapshot.snapshotData.participants as Address[]
+        );
+
+        const virtualVotingResult = this.performVirtualVoting(
+            blockEntries,
+            signerAddress,
+            finalParticipantSet
 
         // Get all blocks sorted by height descending
         const blockEntries = this.storage.blocks.getIterator(
             forkId,
             SortOrder.DESC
         );
 
+        encodedLatestFinalizedState =
+            virtualVotingResult.encodedLatestFinalizedState;
+        encodedLatestCorrectState =
+            virtualVotingResult.encodedLatestCorrectState;
+        virtualVotingBlocks = virtualVotingResult.virtualVotingBlocks;
+
+        // If no finalized state found, use genesis
+        if (!encodedLatestFinalizedState) {
+            encodedLatestFinalizedState =
+                this.storage.getGenesisStateMachineState(forkId) as Bytes;
+        }
+
+        // If no correct state found, use genesis
+        if (!encodedLatestCorrectState) {
+            encodedLatestCorrectState =
+                this.storage.getGenesisStateMachineState(forkId) as Bytes;
+        }
+
+        return {
+            encodedLatestFinalizedState,
+            encodedLatestCorrectState,
+            virtualVotingBlocks,
+            milestoneProofs,
+            milestoneSnapshots
+        };
+    }
+
+    /**
+     * Build a milestone proof for a given set of blocks and participant set
+     */
+    private buildMilestoneProof(
+        blocks: any[],
+        participantSet: Set<Address>
+    ): MilestoneProofStruct | null {
+        if (blocks.length === 0) {
+            return null;
+        }
+
+        const blockConfirmations: BlockConfirmationStruct[] = [];
+        let requiredSignatures = new Set<Address>(participantSet);
```

Comment:
these are not signatures, these are addresses/participants -> so better rename to `requiredParticipants` or similar